### PR TITLE
Enhance creation of bean instances

### PR DESCRIPTION
**Pull request by NazarSmith228**
- Enhance `BeanDefinition` interface
- Enhance `BeanInitializer` - smart creation of bean instances using recursion
- Minor refactoring of `ConfigBasedBeanDefinition`
- Adjust tests

### CODE REVIEW

File: src/main/java/com/bobocode/hoverla/bring/context/BeanDefinition.java [hunk 1]
```diff
@@ -1,42 +1,65 @@
...
+     * Handy method to check whether a bean instance has already been created from current {@link BeanDefinition}.
      *
-     * <p>Requires bean definition which are treated as dependencies to be passed as arguments.</p>
-     * <p>If arguments don't match the actual required dependencies, the exception will be thrown.</p>
+     * @return {@code true} if bean was instantiated, {@code false} otherwise.
+     */
+    boolean isInstantiated();
+
+    /**
+     * Instantiates a bean from current {@link BeanDefinition}using all necessary information encapsulated within.
```
Comment (DanielChanger) [inline src/main/java/com/bobocode/hoverla/bring/context/BeanDefinition.java]: Space is missing here
`{@link BeanDefinition}using`

File: src/main/java/com/bobocode/hoverla/bring/context/BeanInitializer.java [hunk 2]
```diff
@@ -3,40 +3,64 @@
...
 
-        for (BeanDefinition beanDefinition : beanDefinitions) {
-            log.trace("Resolving dependencies for bean definition: {} - {}", beanDefinition.name(), beanDefinition.type());
-            BeanDefinition[] beanDependencies = beanDefinition.dependencies()
-                    .entrySet()
-                    .stream()
-                    .map(keyPair -> beanDefinitionsTable.get(keyPair.getKey(), keyPair.getValue()))
-                    .toArray(BeanDefinition[]::new);
+        BeanDefinition[] beanDependencies = getBeanDependencies(definitionToInitialize, beanDefinitionsTable);
+        int numberOfDependencies = ArrayUtils.getLength(beanDependencies);
```
Comment (DanielChanger) [inline src/main/java/com/bobocode/hoverla/bring/context/BeanInitializer.java:42]: Why not to use just beanDependencies.length, what it gives us? :)

File: src/main/java/com/bobocode/hoverla/bring/context/ConfigBasedBeanDefinition.java [hunk 3]
```diff
@@ -53,120 +53,141 @@ public ConfigBasedBeanDefinition(Object configInstance, Method beanMethod) {
...
         } catch (Exception e) {
             throw new BeanInstanceCreationException("'%s' bean can't be instantiated".formatted(name), e);
         }
     }
 
-    private Object getInstance(BeanDefinition... dependencies) throws InvocationTargetException, IllegalAccessException {
-        Object[] args = Arrays.stream(dependencies)
-                              .map(BeanDefinition::instance)
-                              .toArray();
+    private Object doCreateInstance(BeanDefinition... dependencies) throws InvocationTargetException, IllegalAccessException {
```
Comment (DanielChanger) [inline src/main/java/com/bobocode/hoverla/bring/context/ConfigBasedBeanDefinition.java:169]: Could you please add a test that would make sure it sorts arguments in the expected order?

**Review by DanielChanger — CHANGES_REQUESTED** [review]

File: src/main/java/com/bobocode/hoverla/bring/context/BeanDefinition.java (hunk 1 above)
Comment (NazarSmith228) [reply src/main/java/com/bobocode/hoverla/bring/context/BeanDefinition.java]: good catch

File: src/main/java/com/bobocode/hoverla/bring/context/BeanInitializer.java (hunk 2 above)
Comment (NazarSmith228) [reply src/main/java/com/bobocode/hoverla/bring/context/BeanInitializer.java:42]: Nothing actually) Perhaps a lil bit better in terms of readability...  No hurt for performance) I can change it to usual `.length` if you wish

File: src/main/java/com/bobocode/hoverla/bring/context/ConfigBasedBeanDefinition.java (hunk 3 above)
Comment (NazarSmith228) [reply src/main/java/com/bobocode/hoverla/bring/context/ConfigBasedBeanDefinition.java:169]: will do

**Review by DanielChanger — APPROVED** [review]
LGTM

**Pull request merged**